planex/extract.py

```diff
@@ -0,0 +1,184 @@
+            guard_patch = match.group(2)[3:]
+
+            if gtype == '+' and guard != guard_patch:
+                continue
+            if gtype == '-' and guard == guard_patch:
+                continue
+
+        yield match.group(1)
+
+
+def rewrite_spec(spec_in, spec_out, patches, patchnum):
+    """
+    Expand a patchqueue as a sequence of patches in a spec file
+    """
+    done = False
+    with open(spec_in) as fh_in:
+        with open(spec_out, 'w') as fh_out:
+            for line in fh_in:
+                if not done and line.upper().startswith('SOURCE'):
+                    for patch in patches:
+                        patchnum += 1
+                        fh_out.write("Patch%d: %s\n" % (patchnum, patch))
+                    done = True
+                fh_out.write(line)
+
+
+def expand_patchqueue(args, tar, seriesfile):
+    """
+    Create a list of patches from a patchqueue and update the spec file
+    """
+    # Build a list of patches
+    try:
+        series = tar.extractfile(seriesfile)
+        patches = list(parse_patchseries(series))
+
+    finally:
+        series.close()
+
+    spec = planex.spec.Spec(args.output, topdir=args.topdir,
+                            check_package_name=args.check_package_names)
+    patchnum = spec.highest_patch()
+
+    # Rewrite the spec file to include the patches
+    rewrite_spec(args.output, args.output+".new", patches, patchnum)
+
+    # Switch extracted spec file with new specfile
+    os.rename(args.output, args.output+".old")
+    os.rename(args.output+".new", args.output)
+
+
+def parse_args_or_exit(argv=None):
+    """
+    Parse command line options
+    """
+    parser = argparse.ArgumentParser(description="Extract package sources")
+    add_common_parser_options(parser)
+    parser.add_argument("tarball", help="Tarball")
+    parser.add_argument("-l", "--link", help="Link file")
+    parser.add_argument("-o", "--output", metavar="SPEC",
+                        help="Output spec file")
+    parser.add_argument("-t", "--topdir", metavar="DIR", default=None,
+                        help="Set rpmbuild toplevel directory")
+    parser.add_argument("--no-package-name-check", dest="check_package_names",
+                        action="store_false", default=True,
+                        help="Don't check that package name matches spec "
+                        "file name")
+    argcomplete.autocomplete(parser)
+    return parser.parse_args(argv)
+
+
+def main(argv):
+    """
+    Main function.  Fetch sources directly or via a link file.
+    """
+    setup_sigint_handler()
+    args = parse_args_or_exit(argv)
+    setup_logging(args)
+
+    try:
+        with open(args.link) as fileh:
+            link = json.load(fileh)
+
+    except IOError as exn:
+        # IO error loading JSON file
+        sys.exit("%s: %s: %s" %
+                 (sys.argv[0], exn.strerror, exn.filename))
+
+    # Extract the spec file
+    tar = tarfile.open(args.tarball)
```

Comment:
The Tarfile object can be used as a context manager.   Please change this to 'with tarfile.open(...) as tar'